src/coreclr/vm/pendingload.h

```diff
@@ -17,248 +17,404 @@
+            {
+                TypeKey typeKey = GetTypeKey();
+                ClassLoader::ThrowTypeLoadException(&typeKey,
+                                                    IDS_CLASSLOAD_GENERAL);
+
+            }
+            else
+                EX_THROW(EEMessageException, (m_hrResult));
         }
-        else
-            EX_THROW(EEMessageException, (m_hrResult));
-    }
 
-    void SetException(Exception *pException)
-    {
-        CONTRACTL
+        void SetException(Exception *pException)
         {
-              NOTHROW;
-              PRECONDITION(HasLock());
-              PRECONDITION(m_pException == NULL);
-              PRECONDITION(m_dwWaitCount > 0);
+            CONTRACTL
+            {
+                NOTHROW;
+                PRECONDITION(HasLock());
+                PRECONDITION(m_pException == NULL);
+                PRECONDITION(m_dwWaitCount > 0);
+            }
+            CONTRACTL_END;
+
+            m_typeHandle = TypeHandle();
+            m_hrResult = COR_E_TYPELOAD;
+
+            // we don't care if this fails
+            // we already know the HRESULT so if we can't store
+            // the details - so be it
+            EX_TRY
+            {
+                FAULT_NOT_FATAL();
+                m_pException = pException->Clone();
+            }
+            EX_CATCH
+            {
+                m_pException=NULL;
+            }
+            EX_END_CATCH(SwallowAllExceptions);
         }
-        CONTRACTL_END;
 
-        m_typeHandle = TypeHandle();
-        m_hrResult = COR_E_TYPELOAD;
+        void SetResult(TypeHandle typeHnd)
+        {
+            CONTRACTL
+            {
+                NOTHROW;
+                PRECONDITION(HasLock());
+                PRECONDITION(m_pException == NULL);
+                PRECONDITION(m_dwWaitCount > 0);
+            }
+            CONTRACTL_END;
+
+            m_typeHandle = typeHnd;
+        }
 
-        // we don't care if this fails
-        // we already know the HRESULT so if we can't store
-        // the details - so be it
-        EX_TRY
+        void UnblockWaiters()
         {
-            FAULT_NOT_FATAL();
-            m_pException = pException->Clone();
+            CONTRACTL
+            {
+                NOTHROW;
+                PRECONDITION(HasLock());
+                PRECONDITION(m_dwWaitCount > 0);
+            }
+            CONTRACTL_END;
+
+            _ASSERTE(m_fLockAcquired);
+            m_Crst.Leave();
+            m_fLockAcquired = FALSE;
         }
-        EX_CATCH
+    #endif //DACCESS_COMPILE
+
+        TypeKey& GetTypeKey()
         {
-            m_pException=NULL;
+            LIMITED_METHOD_CONTRACT;
+            return m_typeKey;
         }
-        EX_END_CATCH(SwallowAllExceptions);
-    }
 
-    void SetResult(TypeHandle typeHnd)
-    {
-        CONTRACTL
+        // This is only safe to call on an AddRef'd PendingTypeLoadEntry which has had DelayForProgress() called on it
```

Comment:
Seems like an opportunity for an assert.

Reply:
Yes, but asserts with distributed thread local state are a real headache to write. Instead, I'll adjust `DelayForProgress` to return the `m_typeHandle` when appropriate.